### PR TITLE
feat: add rekognition and s3 interactions to lambda

### PR DESCRIPTION
**Pull request by tsukidust**
template.yaml:
- Added S3 Resource
- Increased lambda timeout from default to 10

app.js:
- Implemented Rekognition detectLabels
- Implemented uploading image to S3
- Passed values to DynamoDB

### CODE REVIEW

File: lambda/src/app.js [hunk 1]
```diff
@@ -49,47 +55,37 @@ exports.lambdaHandler = async (event, context) => {
...
-        country: { S: event.headers["CloudFront-Viewer-Country"] },
-        // TODO any other useful info that can be gleaned?
-      };
+      if (body.image) {
+        const key = id + "." + getFileType(body.image);
 
-      removeNullFields(item);
-      console.log(item);
+        await uploadImageToS3(bucketName, key, imageBuffer);
+        imagePath = getS3ObjectPath(bucketName, key, awsRegion);
```
Comment (sohrab-) [inline lambda/src/app.js:85]: Can you use this path to retrieve the image?

File: lambda/src/app.js [hunk 2]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
...
+async function rekognizeLabels(imageBytes) {
+  const params = {
+    Image: {
+      Bytes: imageBytes,
+    },
+    MaxLabels: 3,
+    MinConfidence: 80,
+  };
+
+  return rekognition.detectLabels(params).promise();
```
Comment (sohrab-) [inline lambda/src/app.js:120]: There was a thing to check if it's sfw, wasn't it?

File: lambda/src/app.js [hunk 3]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
     };
   }
 };
+
+async function rekognizeLabels(imageBytes) {
```
Comment (sohrab-) [inline lambda/src/app.js:111]: We're mixing arrow notation and `function` functions in this file...

File: lambda/src/app.js [hunk 4]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
...
+function getFileType(encodedFile) {
+  return encodedFile.split("data:image/").pop().split(";base64,")[0];
+}
+
+function getImageBuffer(encodedFile) {
+  const base64Image = encodedFile.split("base64,")[1];
+  return Buffer.from(base64Image, "base64");
+}
+
+function getS3ObjectPath(bucket, key, region) {
```
Comment (sohrab-) [inline lambda/src/app.js:171]: This already has access to the region, not sure why passing it in. Could just have the entire URL prefix as the constant. This function doesn't add much.

File: lambda/src/app.js [hunk 5]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
...
+    },
+    tabletClient: {
+      BOOL: event.headers["CloudFront-Is-Tablet-Viewer"] === true,
+    },
+    country: { S: event.headers["CloudFront-Viewer-Country"] },
+    // TODO any other useful info that can be gleaned?
+  };
+
+  removeNullFields(item);
+  console.log(item);
```
Comment (sohrab-) [inline lambda/src/app.js:157]: Remove

File: lambda/src/app.js [hunk 6]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
...
+}
+
+async function insertIntoDynamo(event, id, pet, imagePath) {
+  const item = {
+    id: { S: id },
+    pet: { S: pet },
+    imagePath: { S: imagePath },
+    userAgent: { S: event.headers["User-Agent"] },
+    desktopClient: {
+      BOOL: event.headers["CloudFront-Is-Desktop-Viewer"] === true,
```
Comment (sohrab-) [inline lambda/src/app.js:141]: I realised these are supposed to be `"true"`, which is why we're using equals

File: lambda/src/app.js [hunk 7]
```diff
@@ -111,3 +107,67 @@ exports.lambdaHandler = async (event, context) => {
...
+      Bytes: imageBytes,
+    },
+    MaxLabels: 3,
+    MinConfidence: 80,
+  };
+
+  return rekognition.detectLabels(params).promise();
+}
+
+async function uploadImageToS3(bucket, key, buffer) {
```
Comment (sohrab-) [inline lambda/src/app.js:123]: already got access to the bucket name...

File: lambda/src/app.js (hunk 3 above)
Comment (halfbakedsneed) [reply lambda/src/app.js:111]: Which company's style guide do we want to adhere to?

- Airbnb says [use named functions at the top-level of files](https://github.com/airbnb/javascript#functions--declarations)
- Google implies either is fine, [but seems to prefer arrow functions](https://google.github.io/styleguide/jsguide.html#features-functions-top-level-functions)

File: lambda/src/app.js (hunk 1 above)
Comment (tsukidust) [reply lambda/src/app.js]: changed to put key (imageFileName) into dynamo

**Pull request merged**